Source/DafnyCore/Verifier/ProofObligationDescription.cs

```diff
@@ -308,6 +308,45 @@ public class AssertStatement : ProofObligationDescription {
   }
 }
 
+// The Boogie version does not support custom error messages yet
+public class RequiresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "the precondition always holds"
+      : $"error is impossible: {customErrMsg}";
+
+  public override string FailureDescription =>
+    customErrMsg ?? "a precondition could not be proven";
+
+  public override string ShortDescription => "requires";
+
+  public RequiresDescription([CanBeNull] string customErrMsg = null) {
+    this.customErrMsg = customErrMsg;
+  }
+}
+
+// The Boogie version does not support custom error messages yet
+public class EnsuresDescription : ProofObligationDescription {
+  private readonly string customErrMsg;
+  public override string SuccessDescription =>
+    customErrMsg is null
+      ? "this postcondition holds"
+      : $"error is impossible: {customErrMsg}";
+
+  public override string FailureDescription =>
+    customErrMsg ?? "this postcondition might not hold on a return path";
+
+  public string FailureAtPathDescription =>
```

Comment:
I find the concept of having an `at path` and a regular failure description not immediately intuitive.

Conceptually I view the verification failure as a stack trace. For example in:

```
predicate P(x : int) {
  true
}

predicate Q(x : int) {
  x == 2 // Stack(0)
}
method Foo() return (x: int)
  ensures O(x) && Q(x) // Stack(1)
{
  x := 1;
  assert Foo.ensures() // Stack(2)
  return;
}
```

I think in a previous discussion about squiggly lines, you explained that any part of the stack trace that falls outside of the method definition should not be taken into account, so for failing postconditions that only leaves the bottom 2 stack elements as relevant, and then I see it makes sense to have a failure description for both of those.

Feel free to ignore this comment, just me doing rubber duck analysis.

Reply:
Thanks for rubber duck analysis :-) 
Your metaphor with stack trace is quite interesting, worth exploring from UX's perspective. And I think this PR is exploring it that way. Let me explain you why in the next paragraph.

The existence of "nested expressions" here is due only to the condition splitting at the Dafny level, that is, if we want to prove "A && B", it will prove "A" on one side and "A ==> B" (or just "B" where A is assumed actually) on the other side. If we want to prove `P(x)`, we will prove each conjunct of `P(x)` -- that's also why sometimes it's worth making predicate opaque, because if `P(x)` has 100 conjuncts, proving `P(y)` when `x == y` can still be very costly -
This is why I chose to replace "Could not prove" (which was technically correct, If you can't prove `A`, you can't prove `A && B`) by "Inside", to **better match the metaphor of the stack**.

> I think in a previous discussion about squiggly lines, you explained that any part of the stack trace that falls outside of the method definition should not be taken into account,

My point was that they should not be _underlined permanently_ as errors because they are relevant only in the context of the failed assertion.